### ISSUE TITLE
Handle large datasets efficienlty

### ISSUE THREAD

**Issue by dalonsoa** [ImperialCollegeLondon/virtual_rainforest#582]
- Some models are going to require data at much higher temporal resolution than the wider model update tick. An example here is sub-daily or daily inputs to the Abiotic model.
- The input data files for this use case can be very large – not something we really want to ingest into the Data object at model startup and try and store in RAM.
- So, where do we store this kind of data, and is there a way to lazily load the data as required.  This might be something that `dask` is well-suited to as this handles lazy loading of chunked data.

**Comment by dalonsoa** [comment]
@vgro , we will need an example simulation with, at least, one BIG file and some indication to where it is used, so we can explore how to best handle that memory wise.

**Comment by alexdewar** [comment]
@vgro Do you happen to have a big file like this lying around? No pressure -- I've got lots to be getting on elsewhere -- but I won't be able to start on this until there's some data for me to work with, so if you do have a chance to look at it over the next few weeks, that'd be great.

**Comment by vgro** [comment]
@alexdewar  I'm terribly sorry that I haven't replied to this, I never received an email about the issue and we haven't checked the issues systematically in a while. I have a few urgent tasks this week, I'll try and get something for you by the end of next week or so

**Comment by alexdewar** [comment]
Nw @vgro. If it had been really urgent I'd have sent an email... Whenever you can send it through is fine.

**Comment by vgro** [comment]
> Nw [@vgro](https://github.com/vgro). If it had been really urgent I'd have sent an email... Whenever you can send it through is fine.

If you want to run a simulation, you would probably need all the input data to have the same dimensions? For example you would need the climate data to have the same spatial extent and time steps? Or would it be enough to provide one variable, say precipitation?

**Comment by alexdewar** [comment]
Ideally it would have the same dimensions. I haven't looked into it enough to know exactly what I'd need, but big files with somewhat realistic input data should do the trick. Don't spend too long on this -- feel free to just send it through once you've got something and I can let you know if I need anything different.

**Comment by vgro** [comment]
> Ideally it would have the same dimensions. I haven't looked into it enough to know exactly what I'd need, but big files with somewhat realistic input data should do the trick. Don't spend too long on this -- feel free to just send it through once you've got something and I can let you know if I need anything different.

Hi @alexdewar I think for the first tests (all models have the same time step of one month) it would be the easiest if you took the example data that is part of the VE and repeated the input data with time index (`air_temperature_ref`, `relative_humidity_ref`, `atmospheric_pressure_ref`, `precipitation`, `atmospheric_co2_ref`, `wind_speed_ref`, `topofcanopy_radiation`) for a few hundred/thousand/million years. That should be easier than me generating a large dataset and trying to send it to you.

For the second point, different temporal resolutions, we will need to think a bit more about how to handle this first. We might need to switch from time index to datetime or seconds since start of simulation or something else. I'll get back to you about that.

**Comment by alexdewar** [comment]
Okey doke. I'll have a go at that. Thanks for the response! I'll keep you posted

**Comment by dalonsoa** [comment]
@vgro I've been looking into this to try to understand how datasets with a time dimension are used in the code and, therefore, how handling big files should be best implemented. I've focused on `air_temperature_ref` just to have something concrete to look at.

One of the places it is used is in the setup of the abiotic model, when [calculating the vapour pressure deficit](https://github.com/ImperialCollegeLondon/virtual_ecosystem/blob/343bbcdf45f5c05145fbe14026132f23b56c98ce/virtual_ecosystem/models/abiotic/abiotic_model.py#L177). The input to this function includes `self.data["air_temperature_ref"]`, so you are passing the WHOLE of the dataset, all points in space and time. Then, within the `calculate_vapour_pressure_deficit` function, one of the first things done is to [transform the temperature to a numpy array](https://github.com/ImperialCollegeLondon/virtual_ecosystem/blob/343bbcdf45f5c05145fbe14026132f23b56c98ce/virtual_ecosystem/models/abiotic_simple/microclimate_simple.py#L220).

The main question that come to my mind is: do you need the whole dataset - all points in time - in this calculation? Because if you do, and it needs to be transformed to a `numpy` array, then the complete dataset will need to be loaded in memory and, therefore, using `dask` or other tooling to lazily handle the data is pointless.

If you do not need the whole of the dataset, then we need to change the code so only the relevant time point is passed to the function and possibly, if needed, update the code within the function so it does not expect a time dimension. This review and update will need to be done for all variables containing a time dimension anywhere in the codebase.

So, what's gonna be? :)

**Comment by vgro** [comment]
> [@vgro](https://github.com/vgro) I've been looking into this to try to understand how datasets with a time dimension are used in the code and, therefore, how handling big files should be best implemented. I've focused on `air_temperature_ref` just to have something concrete to look at.
> 
> One of the places it is used is in the setup of the abiotic model, when [calculating the vapour pressure deficit](https://github.com/ImperialCollegeLondon/virtual_ecosystem/blob/343bbcdf45f5c05145fbe14026132f23b56c98ce/virtual_ecosystem/models/abiotic/abiotic_model.py#L177). The input to this function includes `self.data["air_temperature_ref"]`, so you are passing the WHOLE of the dataset, all points in space and time. Then, within the `calculate_vapour_pressure_deficit` function, one of the first things done is to [transform the temperature to a numpy array](https://github.com/ImperialCollegeLondon/virtual_ecosystem/blob/343bbcdf45f5c05145fbe14026132f23b56c98ce/virtual_ecosystem/models/abiotic_simple/microclimate_simple.py#L220).
> 
> The main question that come to my mind is: do you need the whole dataset - all points in time - in this calculation? Because if you do, and it needs to be transformed to a `numpy` array, then the complete dataset will need to be loaded in memory and, therefore, using `dask` or other tooling to lazily handle the data is pointless.
> 
> If you do not need the whole of the dataset, then we need to change the code so only the relevant time point is passed to the function and possibly, if needed, update the code within the function so it does not expect a time dimension. This review and update will need to be done for all variables containing a time dimension anywhere in the codebase.
> 
> So, what's gonna be? :)

Hi Diego,
I don't need ALL the time steps at once, I just felt at the time that it would be easier to calculate vapour pressure deficit for all timesteps in one go, like a pre-processing step. It could be done at each time step or alternatively be a required variable. I think this applies to all variables with a time dimension

**Comment by dalonsoa** [comment]
Well, I guess there's a trade off. Pre-calculating things will speed up the runtime, but will be very heavy on memory for large files since everything needs to be loaded in full at the start (as well as derived quantities)). Potentially, it will be unacceptable if memory requirements are too large. Doing it one step at a time allows for loading the data lazily, only what is needed when is needed, and have in memory only the values required at any time for the derived quantities, but will make runtime slower. Intermediate options seem cumbersome, to be honest.

**Comment by davidorme** [comment]
I think we want to keep the required input variables as close as possible to the commonly available sources (so relative humidity  is something from climate data sources but VPD typically isn't).

If we then have required _derived_ variables (like RH -> VPD) then I think we have three options:

1. As at present, **preprocess in the model startup**, with attendant memory requirements. Just to put a ballpark on that, if we have 100 years of half hourly data for 10000 cells - which seems in the mission spec - then the data size _per variable_ (assuming 4 bytes per float64 value):

    ```python
    In [550]:  (4 * (10000 * 100 * 365 * 48)) / 1024 ** 3
    Out[550]: 65.26708602905273   # GB
    ```
    That doesn't seem tractable.

2. Lazily load and **calculate only on model updates**, with associated time cost in the conversion, which is where this issue is going right now.
 
3. **Create a preprocessor step**. We've got _required_ variables but then models can also have required _derived_ variables.      I think there are few enough of these cases that we could simply write a preprocessing script to create the derived variables and then _save the outputs to file with a data config_. That config can then be included in a model run to provide the data for any number of repeated simulations and different scenarios.
    
    This could be done more elegantly: each model could declare a dictionary of required derived variables and a function to create the derived values. If those named variables are already present in the data on load they get used, nothing happens, but if they are absent then a preprocessor stage after the data load stage runs each recipe and extends the configuration - that could also save the outputs and a config for re-use.

**Comment by vgro** [comment]
I don't think it would be the end of the world to calculate them in each time step, this affects only very few variables and the function could easily be included in the update routine like any other atmospheric variable. Option 3 seems a bit overkill for this specific purpose.

**Comment by davidorme** [comment]
I agree it's a bigger solution. I do think there is a lot of mileage in _not_ continually recalculating derived variables though. If we are already moving the calculations, it might not be a much bigger step.

But it is a different solution to the the direction @dalonsoa is talking about here for these _specific_ variables.

**Comment by dalonsoa** [comment]
I'm talking about this specific variables because they were mentioned by @vgro above, when initially discussing this issue, but it will be good to come up with a generic solution on how to handle large datasets. 

However, to be honest, I do not think that recalculating derived variables every iteration is such a big deal. Based on your calculation above, we are talking on operating on arrays with 10000 of float elements once at the start of the update method, so it should not be too time consuming. Granted that it would take some time, but I'm not convinced that loading 10000 from disk, which will be the case if we implement the 3rd solution, will be faster.

**Comment by davidorme** [comment]
> Granted that it would take some time, but I'm not convinced that loading 10000 from disk, which will be the case if we implement the 3rd solution, will be faster.

It's faster if (and probably _only_ if) you never use the original source variables: those are the commonly available forms of the data but we always use the derived variable in practice. Then you just have to load the derived data, not load and process the source data. It's a pretty marginal gain though.

I think when we get to repeated analysis of the same inputs (grid searches, getting distributions with stochastic variation) this could make enough of a runtime difference to be useful, but it probably isn't a priority.
